lib/puppet/configurer.rb

```diff
@@ -157,25 +157,19 @@ def prepare_and_retrieve_catalog(cached_catalog, options, query_options)
     query_options[:checksum_type] = @checksum_type.join('.')
 
     # apply passes in ral catalog
-    if options[:catalog]
-      options[:catalog]
-    elsif cached_catalog
-      cached_catalog
-    else
+    catalog = cached_catalog || options[:catalog]
```

Comment:
It looks like we're preferring the cached catalog over `options[:catalog]`. This is opposite of what was happening before. Is that on purpose?

Reply:
Yeah, good question. I thought the cached catalog actually takes precedence, even if `puppet apply` case, but let me double check.

Reply:
In 5.5.x, if `puppet apply --use_cached_catalog -e ...` is specified, then the cached catalog option takes precedence, but is ignored, because the catalog cache terminus is set to nil. For example, here I run the agent, get a new catalog, verify it runs from the cache, then try to use the cached catalog with apply, but it's ignored (as we end up print the "from apply" notify message:

```
$ bx puppet agent -t
Info: Using configured environment 'production'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Retrieving locales
Info: Caching catalog for XXX
Info: Applying configuration version '1576608539'
Notice: From Server
Notice: /Stage[main]/Main/Notify[From Server]/message: defined 'message' as 'From Server'
Notice: Applied catalog in 0.01 seconds
$ bx puppet agent -t --use_cached_catalog
Info: Using cached catalog from environment 'production'
Info: Applying configuration version '1576608539'
Notice: From Server
Notice: /Stage[main]/Main/Notify[From Server]/message: defined 'message' as 'From Server'
Notice: Applied catalog in 0.01 seconds
$ bx puppet apply --use_cached_catalog -e "notify { 'from apply': }"
Notice: Compiled catalog for XXX in environment production in 0.02 seconds
Error: /File[/Users/josh/.puppetlabs/opt/puppet/cache/facts.d]: Could not evaluate: Could not retrieve information from environment production source(s) puppet:///pluginfacts
Error: /File[/Users/josh/.puppetlabs/opt/puppet/cache/lib]: Could not evaluate: Could not retrieve information from environment production source(s) puppet:///plugins
Notice: from apply
Notice: /Stage[main]/Main/Notify[from apply]/message: defined 'message' as 'from apply'
```

However, if I set the `catalog_cache_terminus` to json, then the cached catalog does take precedence, which is 🍌 🤡 

```
$ bx puppet apply --use_cached_catalog -e "notify { 'from apply': }" --catalog_cache_terminus json
Notice: From Server
Notice: /Stage[main]/Main/Notify[From Server]/message: defined 'message' as 'From Server'
```

This PR preserves that behavior:

```
$ bx puppet apply --use_cached_catalog -e "notify { 'from apply': }"
Notice: Compiled catalog for XXX in environment production in 0.02 seconds
Error: /File[/Users/josh/.puppetlabs/opt/puppet/cache/facts.d]: Could not evaluate: Could not retrieve information from environment production source(s) puppet:///pluginfacts
Error: /File[/Users/josh/.puppetlabs/opt/puppet/cache/lib]: Could not evaluate: Could not retrieve information from environment production source(s) puppet:///plugins
Notice: from apply
Notice: /Stage[main]/Main/Notify[from apply]/message: defined 'message' as 'from apply'
Notice: Applied catalog in 0.01 seconds
$ bx puppet apply --use_cached_catalog -e "notify { 'from apply': }" --catalog_cache_terminus json
Notice: From Server
Notice: /Stage[main]/Main/Notify[From Server]/message: defined 'message' as 'From Server'
Notice: Applied catalog in 0.01 seconds
```

But I don't think we have a test for that. Let me add one.

Reply:
Thank you for verifying that. So many good times :)

Reply:
So it turns out the configurer was not correct. I've updated that in f4a6909d8d9aa868687d3504a1d9e9a4f80b54ff. The reason I didn't notice when running `puppet apply --use_cached_catalog -e "..."` is because the application itself loads the cached catalog and passes that in as `options[:catalog]`. So then we happen to get lucky because `options[:catalog]` and `@cached_catalog` are both loaded from the cache, and should be equivalent.